Implement preprocessor argument in text index arguments.

Ergus opened the pull request:
<!--- Disable AI PR formatting assistant: true -->
### Changelog category (leave one):
- New Feature

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):

User can now add a new argument `preprocessor` in text index construction. The argument is an arbitrary expression that transforms each document before tokenization.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

- Parameters:

`preprocessor` :  The preprocessor is an arbitrary expression that transforms each document before tokenization.

- Example use:

```
CREATE TABLE users (
    str String,
    index idx() type text(tokenizer = 'default', preprocessor = lower(str))
ENGINE=MergeTree ORDER BY tuple();

SELECT * FROM users WHERE hasToken(name, 'crash');
```

### Details

The text index construction now accepts a new argument which is intended to be a function.

Ergus commented:
Failing test `00900_long_parquet_load` is already reported as flaky in #88893

EmeraldShift commented:
> The argument expects is an arbitrary expression that transforms each document before tokenization.

Can the expression involve more than one column? For example, `arrayConcat(f(col1), g(col2), ...)`? If so, how does one query the text index for the transformed tokens? Is there a form of `has[Any/All]Token` that doesn't require a column name?

Ergus committed:
Allow null context in actions visitor

When we reconstruct the actions dag from a string we don't have context information and we don't need it unless we use a UFD as preprocessor
function.
This change allows to parse with an ActionsVisitor without context.
* src/Interpreters/Context_fwd.h (tryGetContext) : New function that bypasses the check context!= nullptr;
* src/Interpreters/ActionsVisitor.cpp (ActionsMatcher::visit) : Use the new context getter

Ergus committed:
Add functional preprocessor code.

* src/Storages/MergeTree/MergeTreeIndexText.{h,cpp} (MergeTreePreprocessor) : New class to store and parse preprocessor expressions
(MergeTreeIndexAggregatorText::update) : Here is the key part of this change. Using preprocessor->processColumn attempts to avoid unneeded
copies if there is not preprocessor. To do that the preprocessor class is always constructed even with an empty expression and it checks
internally if a copy is needed or not..

* src/Storages/MergeTree/MergeTreeIndexConditionText.{h,cpp} : Reference the new preprocessor member and use it in all the
[sub]stringToTokens[Like] functions.

Ergus committed:
Add basic tests.

TODO: extend this test

Ergus committed:
Add some documentation

Ergus committed:
CI

Ergus committed:
Improve validation code

Use correct error codes and extra validation step.

rschu1ze commented on tests/queries/0_stateless/02346_text_index_preprocessor.sql:
```diff
@@ -0,0 +1,73 @@
+SET allow_experimental_full_text_index = 1;
+SET use_skip_indexes_on_data_read = 1;
+
+-- Basic test preprocessor clause usage with lower
```
Minor: `Tests the preprocessor argument for tokenizers in the text index definitions.`

rschu1ze commented on tests/queries/0_stateless/02346_text_index_preprocessor.sql:
```diff
@@ -0,0 +1,73 @@
+SELECT count() FROM tab WHERE hasToken(str, 'foo');
+SELECT count() FROM tab WHERE hasToken(str, 'FOO');
+
+SELECT count() FROM tab WHERE hasToken(str, 'BAR');
```
L. 23 - 27 test the same thing as l. 20-21.

Ergus replied:
Not exactly, see the new tests with a different preprocessor to find the differences

rschu1ze commented on tests/queries/0_stateless/02346_text_index_preprocessor.sql:
```diff
@@ -0,0 +1,73 @@
+SET allow_experimental_full_text_index = 1;
```
It would be nice to extend existing test 02346_text_index_creation.sql a bit.

rschu1ze commented on tests/queries/0_stateless/02346_text_index_preprocessor.sql:
```diff
@@ -0,0 +1,73 @@
+(
+    key UInt64,
+    str String,
+    str2 String,
```
The term "postprocessor" is pending the result of our internal poll (alternative proposal: "character_filter").

rschu1ze commented on src/Storages/MergeTree/MergeTreeIndexText.h:
```diff
@@ -297,9 +304,30 @@ struct MergeTreeIndexTextGranuleBuilder
     std::unique_ptr<Arena> arena;
 };
 
+class MergeTreePreprocessor
```
The name is too generic.

--> `MergeTreeTextIndexPreprocessor`.

rschu1ze commented on src/Storages/MergeTree/MergeTreeIndexText.h:
```diff
@@ -297,9 +304,30 @@ struct MergeTreeIndexTextGranuleBuilder
+{
+public:
+    MergeTreePreprocessor(const String & expression, const IndexDescription & index_description);
+    std::pair<ColumnPtr,size_t> processColumn(const ColumnWithTypeAndName & index_column_with_type_and_name, size_t start_row, size_t n_rows) const;
```
I think method `processColumn` deserves a comment. It is not immediately clear what the second returned value (size_t) represents.

Ergus replied:
Solved via a86e3a3e1403a4bcfee6c

rschu1ze commented on src/Storages/MergeTree/MergeTreeIndexText.h:
```diff
@@ -297,9 +304,30 @@ struct MergeTreeIndexTextGranuleBuilder
+public:
+    MergeTreePreprocessor(const String & expression, const IndexDescription & index_description);
+    std::pair<ColumnPtr,size_t> processColumn(const ColumnWithTypeAndName & index_column_with_type_and_name, size_t start_row, size_t n_rows) const;
+    String processString(const String &input) const;
```
Needs a source code comment, "processString" is too generic to understand from the name what it is doing.

rschu1ze commented on docs/en/engines/table-engines/mergetree-family/invertedindexes.md:
````diff
@@ -86,6 +87,56 @@ returns
 +---------------------------------+
 ```
 
+Optional parameter `preprocessor` is intended to be an expression that transforms the input string into another one before
````
(minor: convention in this file is one sentence per line)

rschu1ze commented on docs/en/engines/table-engines/mergetree-family/invertedindexes.md:
```diff
@@ -38,6 +38,7 @@ CREATE TABLE tab
                                 -- Mandatory parameters:
                                 tokenizer = splitByNonAlpha|splitByString(S)|ngrams(N)|array
                                 -- Optional parameters:
+                                [, preprocessor = expression(str)]
```
I think we should re-structure this slightly to be consistent with below description:

```sql
                                -- Mandatory parameters:
                                tokenizer = splitByNonAlpha|splitByString(S)|ngrams(N)|array
                                -- Optional parameters:
                                [, preprocessor = expression(str)]
                                -- Optional advanced parameters
                                [, dictionary_block_size = D]
                                [, dictionary_block_frontcoding_compression = B]
                                [, max_cardinality_for_embedded_postings = M]
[, bloom_filter_false_positive_rate = R]
```

Ergus replied:
Do youy mena add subsections like optional and mandatory?

I thought that, but doing so may break the aesthetics of all the document compared with others

rschu1ze commented on docs/en/engines/table-engines/mergetree-family/invertedindexes.md:
````diff
@@ -86,6 +87,56 @@ returns
 ```
 
+Optional parameter `preprocessor` is intended to be an expression that transforms the input string into another one before
+tokenization. The main purposes for this argument are:
````
`[...] main purpose of this argument is [...]`

rschu1ze commented on docs/en/engines/table-engines/mergetree-family/invertedindexes.md:
```diff
@@ -86,6 +87,56 @@ returns
+2. Cleanup the data (i.e remove HTML tags)
+3. Optimize the index information (i.e remove accents, reduce cardinality limiting input to UTF-8)
+
+The expression's input and output are intended to be of the same types, but they are intended to work if they return any of the supported
```
Dumber version for the masses:

```
The expression must transform an input value of type String, [...] to a value of the same type.

Examples:
- `INDEX idx(col) TYPE text(tokenizer = 'splitByNonAlpha', preprocessor = lower(col))`
- [..] <-- insert 2-3 examples here.

The preprocessor expression must only reference the column (or column expression) on top of which the text index is defined.
It is not allowed to use User Defined Functions [UDFs](sql-reference/functions/udf.md) in the preprocessor expression.
Non-deterministic functions are likewise not allowed.

Functions `hasToken`, `hasAllTokens` and `hasAnyTokens` (<-- please link) use the preprocessor to first transofrm the search term before tokenizing it.

For example:

CREATE TABLE tab
(
    key UInt64,
    str String,
    INDEX idx(str) TYPE text(tokenizer = 'splitByNonAlpha', preprocessor = lower(str))
)
ENGINE = MergeTree
ORDER BY tuple();

SELECT count() FROM tab WHERE hasToken(str, 'Foo');

is equivalent to:

CREATE TABLE tab
(
    key UInt64,
    str String,
    INDEX idx(lower(str)) TYPE text(tokenizer = 'splitByNonAlpha')
)
ENGINE = MergeTree
ORDER BY tuple();

SELECT count() FROM tab WHERE hasToken(str, lower('Foo'));
```

Ergus replied:
Solved via 88b1805f9e31a4f8

rschu1ze commented on src/Storages/MergeTree/MergeTreeIndexText.cpp:
```diff
@@ -1383,4 +1479,108 @@ FieldVector MergeTreeIndexText::parseArgumentsListFromAST(const ASTPtr & argumen
+    chassert(index_description.data_types.size() == 1);
+
+    /// Empty expression still creates a preprocessor with empty actions.
+    if (std::ranges::all_of(expression, isspace))
```
We should have a SQL test for this.

rschu1ze commented on src/Storages/MergeTree/MergeTreeIndexText.cpp:
```diff
@@ -1383,4 +1479,108 @@ FieldVector MergeTreeIndexText::parseArgumentsListFromAST(const ASTPtr & argumen
+
+    { /// Parse and verify expression
+        if (!ParserExpression().parse(token_iterator, expression_ast, expected))
+            throw Exception(ErrorCodes::INCORRECT_QUERY, "Error parsing preprocessor expression");
```
`BAD_ARGUMENTS` seems the better error code here (also in l. 1550)

rschu1ze commented on src/Storages/MergeTree/MergeTreeIndexText.cpp:
```diff
@@ -1383,4 +1479,108 @@ FieldVector MergeTreeIndexText::parseArgumentsListFromAST(const ASTPtr & argumen
+        validatePreprocessorASTExpression(preprocessor_function, identifier_name);
+    }
+
+    // TODO(JAM): Check this
```
It's not clear what the code below is doing. Please add some comments.

rschu1ze commented on src/Storages/MergeTree/MergeTreeIndexText.cpp:
```diff
@@ -1336,32 +1413,51 @@ Tuple parseNamedArgumentFromAST(const ASTFunction * ast_equal_function)
-        result.emplace_back(identifier_arg->name());
+        const ASTFunction * preprocessor_function = arguments->children[1]->as<ASTFunction>();
+        if (preprocessor_function == nullptr)
+            throw Exception(ErrorCodes::INCORRECT_QUERY, "Text index preprocessor argument is intended to be a function");
```
`[...] must be [...]`

rschu1ze commented on src/Storages/MergeTree/MergeTreeIndexText.cpp:
```diff
@@ -1312,10 +1332,67 @@ void textIndexValidator(const IndexDescription & index, bool /*attach*/)
+/// Early preprocessor argument validation.
+/// Maybe we could omit this validation and use only the validate function. But here we do it early and simpler to ensure that what we parse
+/// latter is correct
+void validatePreprocessorASTExpression(const ASTFunction * function, String &identifier_name)
```
If possible, please avoid out argument. They are too indirect.

rschu1ze commented on src/Storages/MergeTree/MergeTreeIndexText.cpp:
```diff
@@ -879,22 +889,21 @@ void MergeTreeIndexAggregatorText::update(const Block & block, size_t * pos, siz
 
-    size_t current_position = *pos;
-    const auto & index_column = block.getByName(index_column_name).column;
+    auto [processed_column, offset] = preprocessor->processColumn(block.getByName(index_column_name), *pos, rows_read);
```
For debugging and readability, it will be nicer if `block.getByName(index_column_name)` is on its own line.

rschu1ze replied:
`processed_column` --> `preprocessed_index_column`

rschu1ze commented on src/Storages/MergeTree/MergeTreeIndexText.cpp:
```diff
@@ -1172,8 +1183,16 @@ MergeTreeIndexPtr textIndexCreator(const IndexDescription & index)
     UInt64 max_cardinality_for_embedded_postings = extractOption<UInt64>(options, ARGUMENT_MAX_CARDINALITY_FOR_EMBEDDED_POSTINGS).value_or(DEFAULT_MAX_CARDINALITY_FOR_EMBEDDED_POSTINGS);
     double bloom_filter_false_positive_rate = extractOption<double>(options, ARGUMENT_BLOOM_FILTER_FALSE_POSITIVE_RATE).value_or(DEFAULT_BLOOM_FILTER_FALSE_POSITIVE_RATE);
 
+    String preprocessor = extractOption<String>(options, ARGUMENT_PREPROCESSOR).value_or("");
```
Please move this before l. 1181.

Ergus replied:
Solved via a86e3a3e1403a4bcfee6c

rschu1ze commented on tests/queries/0_stateless/02346_text_index_preprocessor.sql:
```diff
@@ -0,0 +1,73 @@
+
+SELECT 'Test single tokenizer and preprocessor argument.';
+
+CREATE TABLE tab
```
I think we'll need also tests for indexes on Array and Map columns.

Ergus commented on tests/queries/0_stateless/02346_text_index_preprocessor.sql:
```diff
@@ -0,0 +1,73 @@
+
+-- Basic test preprocessor validation
+SELECT 'Test preprocessor validations.';
+
```
We should disable them? Filter operations are isomorphic operations, so they should produce 1-1 inpout-output right?

Confirm and I will add it

Ergus replied:
Solved via 2d1f2b68b76

CurtizJ commented on src/Storages/MergeTree/MergeTreeIndexText.cpp:
```diff
@@ -1383,4 +1479,108 @@ FieldVector MergeTreeIndexText::parseArgumentsListFromAST(const ASTPtr & argumen
+    ColumnNumbersList aggregation_keys_indexes_list;
+
+    ActionsVisitor::Data visitor_data(
+        nullptr,
```
Unfortunately you cannot pass `nullptr` as context, because sometimes it is still needed for functions:
```sql
ALTER TABLE hackernews_index
    (ADD INDEX idx_text text TYPE text(tokenizer = splitByNonAlpha, preprocessor = translate(text, 'üß', 'us')) GRANULARITY 64)

Query id: adb59fca-4c50-40a3-8cf6-b1f037e76eb0

2025.10.29 23:06:19.916064 [ 88946 ] {adb59fca-4c50-40a3-8cf6-b1f037e76eb0} <Error> executeQuery: Code: 49. DB::Exception: Context has expired: While processing translate(text, 'üß', 'us'). (LOGICAL_ERROR) (version 25.11.1.537) (from 127.0.0.1:37486) (query 5, line 7) (in query: ALTER TABLE hackernews_index ADD INDEX idx_text text TYPE text(tokenizer = splitByNonAlpha, preprocessor = translate(text, 'üß', 'us'));), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x0000000013e53cdf
1. DB::Exception::Exception(String&&, int, String, bool) @ 0x000000000cb2850e
2. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x000000000cb27fc0
3. DB::Exception::Exception<>(int, FormatStringHelperImpl<>) @ 0x000000000cb3836b
4. DB::WithContextImpl<std::shared_ptr<DB::Context const>>::getContext() const @ 0x000000000cb38126
5. DB::ActionsMatcher::visit(DB::ASTLiteral const&, std::shared_ptr<DB::IAST> const&, DB::ActionsMatcher::Data&) @ 0x0000000018a03c71
6. DB::ActionsMatcher::visit(DB::ASTFunction const&, std::shared_ptr<DB::IAST> const&, DB::ActionsMatcher::Data&) @ 0x00000000189fd17f
7. DB::InDepthNodeVisitor<DB::ActionsMatcher, true, false, std::shared_ptr<DB::IAST> const>::doVisit(std::shared_ptr<DB::IAST> const&) @ 0x00000000189f2b15
8. DB::MergeTreePreprocessor::parseExpression(DB::IndexDescription const&, String const&) @ 0x0000000019fb03f2
9. DB::textIndexValidator(DB::IndexDescription const&, bool) @ 0x0000000019faeadd
10. DB::MergeTreeData::checkProperties(DB::StorageInMemoryMetadata const&, DB::StorageInMemoryMetadata const&, bool, bool, bool, bool, std::shared_ptr<DB::Context const>) const @ 0x0000000019e0f6b7
11. DB::MergeTreeData::checkAlterIsPossible(DB::AlterCommands const&, std::shared_ptr<DB::Context const>) const @ 0x0000000019e604e9
12. DB::InterpreterAlterQuery::executeToTable(DB::ASTAlterQuery const&) @ 0x0000000018a6acab
13. DB::InterpreterAlterQuery::execute() @ 0x0000000018a678cd
```

Ergus committed:
Extend the preprocessor tests to check more corner case constructors.

Ergus committed:
Add check for non deterministic preprocessor actions

Also avoid using nullptr as context by accessing global context throw static.

Ergus committed:
Revert "Allow null context in actions visitor"

This reverts commit ae56d6fb8ec1c2622ec7ec2c6bb8aab84e67dd42.

This was actually a bad idea

Ergus committed:
sed 's/MergeTreePreprocessor/MergeTreeIndexTextPreprocessor/g'

Ergus committed:
Improve the documentation.

Ergus committed:
CR1 @rschu1ze

Ergus commented on src/Storages/MergeTree/MergeTreeIndexText.cpp:
```diff
@@ -1336,32 +1413,51 @@ Tuple parseNamedArgumentFromAST(const ASTFunction * ast_equal_function)
-        result.emplace_back(identifier_arg->name());
+        const ASTFunction * preprocessor_function = arguments->children[1]->as<ASTFunction>();
+        if (preprocessor_function == nullptr)
+            throw Exception(ErrorCodes::INCORRECT_QUERY, "Text index preprocessor argument is intended to be a function");
```
Solved via a86e3a3e1403a4bcfee6c

Ergus commented on src/Storages/MergeTree/MergeTreeIndexText.cpp:
```diff
@@ -1383,4 +1479,108 @@ FieldVector MergeTreeIndexText::parseArgumentsListFromAST(const ASTPtr & argumen
+        /// We already made this check during index construction, but "don't trust, verify"
+        const ASTFunction * preprocessor_function = expression_ast->as<ASTFunction>();
+        if (preprocessor_function == nullptr)
+            throw Exception(ErrorCodes::INCORRECT_QUERY, "Text index preprocessor argument is intended to be a function");
```
Solved via a86e3a3e1403a4bcfee6c

Ergus commented on src/Storages/MergeTree/MergeTreeIndexText.h:
```diff
@@ -297,9 +304,30 @@ struct MergeTreeIndexTextGranuleBuilder
+{
+public:
+    MergeTreePreprocessor(const String & expression, const IndexDescription & index_description);
+    std::pair<ColumnPtr,size_t> processColumn(const ColumnWithTypeAndName & index_column_with_type_and_name, size_t start_row, size_t n_rows) const;
```
Solved via a86e3a3e1403a4bcfee6c

Ergus committed:
CI

ahmadov commented on src/Storages/MergeTree/MergeTreeIndexText.cpp:
```diff
@@ -1383,4 +1495,124 @@ FieldVector MergeTreeIndexText::parseArgumentsListFromAST(const ASTPtr & argumen
+
+    /// Construct a visitor to parse the AST to build a DAG
+    ActionsVisitor::Data visitor_data(
+        Context::getGlobalContextInstance(),
```
Recently, I had an issue with the global context that setting changes were not applied. The global context only holds the default values for the settings but does not reflect changes. We need to pass the local context here instead of forwarding the global context.

ahmadov replied:
Yes, I understand the struggle to pass local context here (touching many files is pain). We can keep as it is.

ahmadov commented on src/Storages/MergeTree/MergeTreeIndexText.cpp:
```diff
@@ -1383,4 +1495,124 @@ FieldVector MergeTreeIndexText::parseArgumentsListFromAST(const ASTPtr & argumen
+    chassert(index_description.data_types.size() == 1);
+
+    /// Empty expression still creates a preprocessor with empty actions.
+    if (std::ranges::all_of(expression, isspace))
```
[minor] std::isspace?

ahmadov commented on src/Storages/MergeTree/MergeTreeIndexText.cpp:
```diff
@@ -1336,37 +1415,70 @@ Tuple parseNamedArgumentFromAST(const ASTFunction * ast_equal_function)
+            throw Exception(ErrorCodes::INCORRECT_QUERY, "Text index preprocessor argument must be an expression");
+
+        String identifier_name;
+        validatePreprocessorASTExpression(preprocessor_function, identifier_name);
```
can we make this part generic but not `preprocessor` specific? This validation is kinda part of semantic analysis not the parsing. I'd prefer to keep parsing and semantic analysis steps separately.

ahmadov replied:
Yes, that is the preprocessor semantic analysis but my point is that semantic analysis should be after the parsing process fully completed. Here we are trying to parse and do semantic analysis at the same time. The initial phase of parsing happens from `IndicesDescription.cpp:112` which calls `validatePreprocessorASTExpression` transitively during this phase. Then in `textIndexValidator` we call `validatePreprocessorASTExpression` again (from `MergeTreeIndexTextPreprocessor::parseExpression`).
IMO the first call from `IndicesDescription.cpp:112` should be as simple as possible (just collect identifiers, literals, tuples, etc) and number of arguments can vary.

All other validations should happen at `textIndexValidator`.

BUT it's up to you. This is also fine.

rschu1ze committed:
Minor doc update

rschu1ze committed:
Minor fixups in the test

Ergus committed:
CR @ahmadov

Ergus commented on src/Storages/MergeTree/MergeTreeIndexText.cpp:
```diff
@@ -1285,6 +1306,7 @@ void textIndexValidator(const IndexDescription & index, bool /*attach*/)
 
     /// No validation for max_cardinality_for_embedded_postings.
     extractOption<UInt64>(options, ARGUMENT_MAX_CARDINALITY_FOR_EMBEDDED_POSTINGS);
+    auto preprocessor_str = extractOption<String>(options, ARGUMENT_PREPROCESSOR, false);
```
Solved via a6280e33552bf952bb

rschu1ze commented on src/Storages/MergeTree/MergeTreeIndexConditionText.h:
```diff
@@ -56,7 +59,7 @@ class MergeTreeIndexConditionText final : public IMergeTreeIndexCondition, WithC
     const Block & getHeader() const { return header; }
 
-    /// Create text search query for the function node if it is suitable for optimization.
+    /// Create text search qupreprocessorery for the function node if it is suitable for optimization.
```
qupreprocessorery

Ergus replied:
Solved via  a0f736318e655c6747b5

rschu1ze commented on tests/queries/0_stateless/02346_text_index_preprocessor.sql:
```diff
@@ -0,0 +1,165 @@
+    str String,
+    INDEX idx(str) TYPE text(tokenizer = 'splitByNonAlpha', preprocessor = nonExistingFunction(str))
+)
+ENGINE = MergeTree ORDER BY tuple();   -- { serverError UNKNOWN_FUNCTION }
```
Minor: If the error is from "our" code, would it be possible to throw `BAD_ARGUMENTS`? Same in l. 110. and l. 128.

Ergus replied:
Added UDF test and it works.

rschu1ze commented on src/Storages/MergeTree/MergeTreeIndexText.h:
```diff
@@ -297,9 +304,43 @@ struct MergeTreeIndexTextGranuleBuilder
+
+    /// Processes n_rows rows of the column in index_column_with_type_and_name starting at start_row.
+    /// The transformation is only applied in the range [start_row, start_row + n_rows)
+    /// If the expression is empty this functions is just a pass-throw.
```
`is just a no-op`?

Ergus replied:
Solved via a0f736318e655c6747b5

ahmadov commented on src/Storages/MergeTree/MergeTreeIndexText.cpp:
```diff
@@ -855,11 +864,13 @@ void MergeTreeIndexTextGranuleBuilder::reset()
     , params(std::move(params_))
     , token_extractor(token_extractor_)
     , granule_builder(params, token_extractor_)
+    , preprocessor(preprocessor_)
```
[nit] `std::move(preprocessor_)`?

ahmadov commented on src/Storages/MergeTree/MergeTreeIndexText.cpp:
```diff
@@ -1312,10 +1333,67 @@ void textIndexValidator(const IndexDescription & index, bool /*attach*/)
+/// Early preprocessor argument validation.
+/// Maybe we could omit this validation and use only the validate function. But here we do it early and simpler to ensure that what we parse
+/// latter is correct
+void validatePreprocessorASTExpression(const ASTFunction * function, String &identifier_name)
```
[nit] space between `&` and `identifier_name`

ahmadov commented on src/Storages/MergeTree/MergeTreeIndexText.cpp:
```diff
@@ -1312,10 +1333,67 @@ void textIndexValidator(const IndexDescription & index, bool /*attach*/)
+            if (identifier_name.empty())
+                identifier_name = identifier->name();
+            else if (identifier_name != identifier->name())
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "Text preprocessor function should receive only one identifier, but there is {} and {}",
```
[minor] "Text index preprocessor ..."

ahmadov reviewed (commented):
overall lgtm

Ergus commented on tests/queries/0_stateless/02346_text_index_preprocessor.sql:
```diff
@@ -0,0 +1,165 @@
+(
+    key UInt64,
+    str String,
+    INDEX idx(lower(str)) TYPE text(tokenizer = 'splitByNonAlpha', preprocessor = lower(str))
```
Hi @rschu1ze 

This `CREATE TABLE` fail because we cannot use `idx(lower(str))` and a preprocessor at the same time. When there is a preprocessor we want the idx argument to be only a column identifier.  

0. Using both simultaneously doesn't add value, complicates the semantic and adds extra complexity to the code.
1. We decided to not support both as a design choice to provide simpler/clearer interface to the user.
2. It enforce the use of `hasToken(lower(str)....` which is so confusing that we decided to implement the preprocessor.
3. We only support the `idx(lower(str))` semantic for backward compatibility, and as a legacy feature we don't want to "promote" its use in newer code (that use newer features)

Actually Once the preprocessor will be aout and tested I would recommend to add a deprecation warning for semantic like `idx(lower(str))` for text index.

Ergus committed:
CR2 @ahmadov & @rschu1ze

Ergus commented on src/Storages/MergeTree/MergeTreeIndexConditionText.h:
```diff
@@ -125,6 +128,9 @@ class MergeTreeIndexConditionText final : public IMergeTreeIndexCondition, WithC
     bool use_bloom_filter = true;
     /// If global mode is All, then we can exit analysis earlier if any token is missing in granule.
     TextSearchMode global_search_mode = TextSearchMode::All;
+
```
Solved via  a0f736318e655c6747b5

Ergus commented on tests/queries/0_stateless/02346_text_index_preprocessor.sql:
```diff
@@ -0,0 +1,165 @@
+(
+    key UInt64,
+    str String,
+    INDEX idx(str) TYPE text(tokenizer = 'splitByNonAlpha', preprocessor = rand())
```
Solved via  a0f736318e655c6747b5

Ergus committed:
Merge remote-tracking branch 'principal/master' into fts_preprocessor

Ergus committed:
Improve tests

1. Add more selects to the output to identify failing negative tests
2. Add udf preprocessor functions

Ergus committed:
Small improves in parser code.

Improve the blank check
Add comment about context

Ergus committed:
Merge remote-tracking branch 'principal/master' into fts_preprocessor

Ergus committed:
Don't mention Arrays in preprocessor doc (yet)

ahmadov commented on src/Storages/MergeTree/MergeTreeIndexText.cpp:
```diff
@@ -1306,10 +1327,67 @@ void textIndexValidator(const IndexDescription & index, bool /*attach*/)
+
+        const Names required_columns = expression.getRequiredColumns();
+
+        /// This is expected top no never happen because the `validatePreprocessorASTExpression` already checks that we have a single identifier.
```
`This is expected that never happen`?

Ergus committed:
Add no-parallel tag to preprocessor test

Ergus committed:
Fix typo in a comment CR3: @ahmadov

Ergus committed:
Merge remote-tracking branch 'principal/master' into fts_preprocessor

Ergus committed:
Merge remote-tracking branch 'principal/master' into fts_preprocessor

Ergus committed:
Emmit right error message on array+preprocessor

Ergus committed:
Merge remote-tracking branch 'principal/master' into fts_preprocessor

clickgapai commented:
Hi @Ergus — while reviewing this PR I found the following:

- **Bug (P1):** `hasAnyTokens`/`hasAllTokens` with array needle don't apply preprocessor to array elements — https://github.com/ClickHouse/ClickHouse/issues/104414

Happy to discuss — close anything that's wrong or already addressed.
